sketches essay (#1828)
instantdb/instant: client/www/_posts/count_min_sketch.md

```typescript
console.log('allSketch beetle', check(allSketch, stem('beetle')));
```

We'll see 102!

```bash
allSketch beetle 102
```

A bit over, but not so bad. [^8]

If you're curious, try out different sizes and see what you get:

<sketch-demo demo="configurable-try-sketch"></sketch-demo>

# A breather to celebrate

Congratulations! You just build out a Count-Min Sketch from scratch, and used it on Wodehouse.

If you'd like to see the full code example, I put this up in it's entirety on <a href="https://github.com/instantdb/count-min-sketch" target="_blank">GitHub</a>.

Hope you had a lot of fun 🙂. If you're still curious there's more to learn here. I present you...2 bonus sections!

# Bonus 1: Probabilities

When we created our sketch for Wodehouse, we chose some seemingly random numbers: 4000 columns and 7 rows. Is there a method to this madness?

Absolutely. We can use some math to help set bounds around our estimations.

## Error Rate & Confidence

There are two numbers we can play with:

1. The **errorRate** tells us how far off we expect our estimation to be
2. The **confidence** tells us how likely it is that we are actually within our estimation.

> Let's make them concrete. The full text for Wodehouse is about 3.7 million words long (not unique words, here we are counting every occurence).

Reviewer: ```suggestion
Let's make them concrete. The full text for Wodehouse is about 3.7 million words long (not unique words, here we are counting every occurrence).
```

So if we write our our expectation, it would be:

$$
expectedNoise_{word} = \frac{totalWords - actualCount_{word}}{columns}
$$

### Simplifying Noise

If you think about, do we really _need_ to subtract the $actualCount_{word}$? It's going to be such a small part of the total anyways.

We can simplify this formula by getting more conservative about what we promise. Let's just say that the the expected noise is _smaller_ than this:

$$
expectedNoise_{word} <= \frac{totalWords}{columns}
$$

Pretty cool. Now we have a simple relation for our expected noise!

### Help from Markov

But an expected value for noise isn't useful yet. It just gives us an average. What we want is the _probability_ that something is below some `maximumOvercount`.

That's where **Markov's Inequality** [^9] comes in. Markov's Inequality is a proof about random variables that says:

> For any non-negative random variable, the probability that something exceeds n times its expected value is at most 1/n.

To get concrete, if we plug in `n = 2` [^14] to Markov's Inequality, we get:

> > The probability that somethings exceeds 2 times it's expected value, is at most 1 / 2

Reviewer: ```suggestion
> The probability that something exceeds 2 times its expected value, is at most 1 / 2
```

> Well, our noise is a non-negative random variable [^12]. And we have it's expected value. If we use Markov's Inequality, we'll get a real probability that we can use!

Reviewer: ```suggestion
Well, our noise is a non-negative random variable [^12]. And we have its expected value. If we use Markov's Inequality, we'll get a real probability that we can use!
```
